### PR TITLE
⚠️ Refactor source/handler/predicate packages to remove dep injection

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vince@prigna.com>

### CODE REVIEW

File: hack/test-all.sh
Comment (vincepri) [inline hack/test-all.sh:28]: These were actually useful to catch errors right away, and have a bit more verbosity when tests are running, we should probably keep them

File: pkg/builder/controller_test.go
Comment (vincepri) [inline pkg/builder/controller_test.go:160]: This is now the new experience on how to create a controller in a manager. Over time, the builder has increased the options we're passing in, which is ok, but we've also had to deal with all of the dependency injection stuff which isn't pretty at all.

This UX might be a stopgap, but it's actually nicer to be explicit where a source is getting its cache or any other function. In a follow-up PR we might want to explore easier ways to plow through this information when we're clearly building a controller in a Manager, and make the source aware that it can access whatever it needs through the manager.

For now, we can keep this as a first step breaking change and iterate on it as we see fit. The clarity might come handy in the future when we reason through the codebase.

File: pkg/builder/controller.go
Comment (vincepri) [reply pkg/builder/controller.go:258]: I'm with you that this isn't ideal, although the UX for building sources with PartialObjectMetadata is pretty poor if we don't do this for our users. That would basically mean creating an PartialObjectMetadata, figure out the GVK, potentially using the scheme directly, which can cause errors. We should probably keep this as-is for now, and figure out a different way later

File: pkg/source/source.go
Comment (vincepri) [reply pkg/source/source.go:73]: done, ptal

File: pkg/source/source.go
Comment (vincepri) [reply pkg/source/source.go]: We can but the EnqueueRequestForOwner counterpart for example takes any required parameters before taking the object, wdyt?

File: pkg/predicate/predicate.go
Comment (vincepri) [reply pkg/predicate/predicate.go:245]: I went ahead and removed most of the injection code except for the ones that touches webhooks, which we should handle in a separate PR

File: pkg/builder/controller.go
Comment (sbueringer) [inline pkg/builder/controller.go:257]: nit: Comment seems to be outdated

Comment (sbueringer) [reply pkg/builder/controller.go:257]: Q: is it correct to check for SyncingSource and then pass the src into KindAsPartialMetadata where we check for kind?

Could it be a SyncingSource which is not kind? ("Users may build their own Source implementations.")

File: pkg/cluster/cluster_test.go
Comment (sbueringer) [inline pkg/cluster/cluster_test.go:175]: Do we still need the injectable? Seems unused (except l.172+173)

File: pkg/handler/enqueue_owner.go
Comment (sbueringer) [inline pkg/handler/enqueue_owner.go:49]:
```suggestion
func EnqueueRequestForOwner(scheme *runtime.Scheme, mapper meta.RESTMapper, ownerType client.Object, opts ...OwnerOption) EventHandler {
```

Just wondering if the parameter should be called ownerType as well

File: pkg/source/source.go
Comment (sbueringer) [inline pkg/source/source.go:96]: Should this be objType? (just wondering why obj here vs ownerType elsewhere, fells like we could keep it consistent)

File: pkg/source/source.go
Comment (sbueringer) [inline pkg/source/source.go:207]:
```suggestion
	// stop is to end ongoing goroutine, and close the channels
```
nit: given the field name is lower case?

File: pkg/source/source_test.go
Comment (sbueringer) [inline pkg/source/source_test.go:181]:
```suggestion
		It("should return an error from Start if cache was not provided", func() {
```
?

File: pkg/builder/controller.go
Comment (sbueringer) [reply pkg/builder/controller.go]: I have a related comment below. not sure if we can actually do this as is https://github.com/kubernetes-sigs/controller-runtime/pull/2120#discussion_r1068235965

File: pkg/builder/controller.go
Comment (vincepri) [reply pkg/builder/controller.go:257]: I agree it's not ideal. Originally I had a `KindSource` instead with `AsPartialMetadata()` method required. These builders need a lot of rework honestly, the UX also isn't ideal with modifiers in place. I'm happy to go back to the interface if we prefer that.

File: pkg/cluster/cluster_test.go
Comment (vincepri) [reply pkg/cluster/cluster_test.go:175]: Thanks! Leftovers :)

File: pkg/builder/controller.go
Comment (vincepri) [reply pkg/builder/controller.go:257]: PTAL, added a new interface called `TypedSource` which lets users provide their own if needed

Comment (sbueringer) [reply pkg/builder/controller.go:257]: lgtm

**Comment by sbueringer** [comment]
lgtm pending squash from my side

**Comment by vincepri** [comment]
@sbueringer Let's keep the commits separate for ease of history later, wdyt?

**Comment by sbueringer** [comment]
> @sbueringer Let's keep the commits separate for ease of history later, wdyt?

Oh sure. Didn't look if they were intentionally separated to be honest 😀

/lgtm

/hold
So Alvaro can take another look as well

File: pkg/builder/controller.go
Comment (alvaroaleman) [inline pkg/builder/controller.go:205]: What I meant with the other comment is that it would be nicer to have a `source.NewMetadataKind(obj client.Object, scheme *runtime.scheme)`, right now it requires to first build a `kind` source and then convert it, which is a bit awkward IMHO

File: pkg/controller/controller.go
Comment (alvaroaleman) [inline pkg/controller/controller.go:143]: Pretty sure this will break people, as Kubebuilder at least used to scaffold in a way that this ended up being done, not sure if that is still the case today. You sure you want to just remove this instead of warn for one release if any of the things end up getting injected?

File: pkg/source/source.go
Comment (alvaroaleman) [inline pkg/source/source.go:71]: Maybe `KindSource` to make clear that the two are related?

File: pkg/source/source.go
Comment (alvaroaleman) [inline pkg/source/source.go:180]: Do you think `ks.started` is worth protecting with a mutex?

**Review by alvaroaleman — COMMENTED** [review]
Sorry it took so long, just some nits.

File: pkg/controller/controller.go
Comment (vincepri) [reply pkg/controller/controller.go:143]: Yes, these fields have been deprecated since 0.10; it's better to not linger the support for dependency injection longer. We can do a couple of things to avoid breaking, including giving a heads up to folks in the mailing list.

File: pkg/builder/controller.go
Comment (vincepri) [reply pkg/builder/controller.go]: Agreed, we need more thinking on how we handle the Metadata projection stuff, reverted some changes and instead moved the Kind object in an internal package

**Review by alvaroaleman — APPROVED** [review]

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**
